### PR TITLE
Replace morgan with bunyan to produce logs in JSON format

### PR DESCRIPTION
**Pull request by scinos**
### Background

In order to improve log coverage on the server, we need to send the logs to Elasticsearch. But ES only consumes logs in JSON format, so we need to add more structure to them.

### Changes

* Replace `morgan` with `bunyan` to produce access logs in JSON format. I tried to log the same info than the previous formats (`dev` and `combined` for `development` and `production`, respectively) by leveraging `morgan` parser functions.

### Testing instructions

* Checkout this branch and run `yarn start`. You should see logs in structured (*) format:
    ```
    12:46:43.279Z  INFO calypso:  (method=GET, status=304, duration=0.363)
        url: /calypso/evergreen/async-load-blocks-support-article-dialog.js
    12:46:43.281Z  INFO calypso:  (method=GET, status=304, duration=0.437)
        url: /calypso/evergreen/async-load-blocks-legal-updates-banner.css
    12:46:43.293Z  INFO calypso:  (method=GET, status=304, duration=0.36)
    ```
    (*) when running `yarn start` I'm using `bunyan` CLI tool to prettify the logs, that's why they they don't look like JSON.

* Stop the server and run it with `NODE_ENV=production yarn start`, you should see more fields in the logs:
    ```
    12:56:04.229Z  INFO calypso:  (method=GET, status=404, length=4150, url=/__webpack_hmr, duration=740.129, remoteAddr=::1, httpVersion=1.1, referrer=http://calypso.localhost:3000/)
        userAgent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36
    ```

* Stop the server and run it again with `node build/server.js`. Now you should see a pure JSON log.
    ```
    {"name":"calypso","hostname":"sergio.lan","pid":59465,"level":30,"method":"GET","status":"404","length":"4150","url":"/calypso/evergreen/vendors~entry-login~entry-main.js","duration":12.282,"remoteAddr":"::1","httpVersion":"1.1","referrer":"http://calypso.localhost:3000/","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36","msg":"","time":"2020-09-10T12:56:53.743Z","v":0}
    ```

### Not covered

This PR does not cover transforming other logs into JSON format (eg: usages of console.log still produce raw text lines). That will be addressed in a future PR to keep things simpler.

### CODE REVIEW

File: client/server/middleware/logger.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import { getLogger } from 'server/lib/logger';
+
+const NS_TO_MS = 1e-6;
+
+const logRequest = ( req, res, options ) => {
+	const { requestStart } = options;
+
+	// Equivalent to morgan's `dev` format
+	let fields = {
+		method: morgan.method( req ),
```
Comment (jsnajdr) [inline client/server/middleware/logger.js:18]: Do I understand correctly that we no longer use `morgan` as the actual logging middleware for Express.js, but use only its formatting functions?

Comment (scinos) [reply client/server/middleware/logger.js:18]: Correct.

We could even drop `morgan` entirely if we want, I'm not 100% sure we want to log exactly the same fields.

File: client/server/lib/logger/index.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/**
+ * External dependencies
+ */
+import bunyan from 'bunyan';
```
[the remark this answers is not in the document]
Comment (scinos) [reply client/server/lib/logger/index.js:4]: Good catch, will fix.

File: client/server/middleware/logger.js (hunk 1 above)
Comment (jsnajdr) [reply client/server/middleware/logger.js:18]: Most of the `morgan` token functions are fairly trivial, like `method = req => req.method` and the dependency could be eliminated without much effort. Completely optional, it's up to you whether you want to do it in this PR or at all.

File: client/server/middleware/logger.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+		),
+	};
+
+	if ( process.env.NODE_ENV === 'production' ) {
+		// Standard Apache combined log output.
+		// Equivalent to morgan's `combined` format
+		fields = {
+			...fields,
+			remoteAddr: morgan[ 'remote-addr' ]( req ),
+			remoteUser: morgan[ 'remote-user' ]( req ),
```
Comment (jsnajdr) [inline client/server/middleware/logger.js:33]: We can drop the `remote-user` field: it looks at the `Authorization` header and reports the username if `basic` HTTP auth is used. We never use that -- we do either OAuth or session cookies.

File: client/server/middleware/test/logger.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	const current = process.env.NODE_ENV;
+	process.env.NODE_ENV = env;
+
+	afterAll( () => {
+		process.env.NODE_ENV = current;
+	} );
+};
+
+beforeAll( () => {
+	jest.useFakeTimers( 'modern' );
```
Comment (jsnajdr) [inline client/server/middleware/test/logger.js:56]: The tests would be better isolated from each other if this was a `beforeEach` method, resetting the fake timers before each test. The [jest docs](https://jestjs.io/docs/en/timer-mocks) recommend that, too.

TIL about the `modern` implementation :+1: We've had some troubles before with Jest fake timers not patching `Date`, and had to use the Sinon implemenation instead.

File: client/server/middleware/test/logger.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+} );
+
+afterAll( () => {
+	jest.useRealTimers();
```
Comment (jsnajdr) [inline client/server/middleware/test/logger.js:60]: This is probably redundant -- Jest will all the mocks et all automatically at the end of the test file.

File: client/server/middleware/test/logger.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+} );
+
+it( 'Adds a `logger` property to the request', () => {
+	const req = withRequest( {} );
+	const res = withResponse( {} );
+	const next = () => {};
+
+	loggerMiddleware()( req, res, next );
+
+	expect( req.logger ).toEqual( mockLogger );
```
Comment (jsnajdr) [inline client/server/middleware/test/logger.js:70]: This should check if the objects are `===`-identical, i.e., `.toBe()`.

**Review by jsnajdr — APPROVED** [review]
The `devDependency` is the only thing that should really be fixed before merging, the rest are mostly nits.

File: client/server/middleware/logger.js (hunk 3 above)
Comment (scinos) [reply client/server/middleware/logger.js:33]: That's great news. I started to re-implement morgan tokens manually as you suggest in another comment, but then found that the implementation of `remote-user` is not trivial. That's when I pivoted to reuse `morgan`'s tokens.

But if we can drop `remote-user`, I can go back to the original idea of reimplement the same tokens and get rid of `morgan`

Comment (jsnajdr) [reply client/server/middleware/logger.js:33]: Now I realize that if we wanted, we could very easily find the real WP.com username in the cookie headers and log it. The `wordpress_logged_in` cookie has [format](https://github.com/WordPress/WordPress/blob/master/wp-includes/pluggable.php#L761):
```
username|expiration|token|hash
```
where the `username` token is simply the username, e.g., `jsnajdr`.

That works for the cookie auth that the web Calypso uses exclusively.

**Comment by jsnajdr** [comment]
The changes look good, thanks for addressing all the feedback 💟 Confirming my approval ✅

**Pull request merged**